File: src/main/java/com/freya02/botcommands/internal/application/localization/BCLocalizationFunction.java
package com.freya02.botcommands.internal.application.localization;

import com.freya02.botcommands.api.Logging;
import com.freya02.botcommands.api.builder.DebugBuilder;
import com.freya02.botcommands.api.localization.Localization;
import com.freya02.botcommands.api.localization.LocalizationTemplate;
import com.freya02.botcommands.internal.BContextImpl;
import net.dv8tion.jda.api.interactions.DiscordLocale;
import net.dv8tion.jda.api.interactions.commands.localization.LocalizationFunction;
import org.jetbrains.annotations.NotNull;
import org.slf4j.Logger;

import java.util.HashMap;
import java.util.List;
import java.util.Locale;
import java.util.Map;

public class BCLocalizationFunction implements LocalizationFunction {
	private static final Logger LOGGER = Logging.getLogger();
	private final Map<String, List<Locale>> baseNameToLocalesMap;

	public BCLocalizationFunction(BContextImpl context) {
		baseNameToLocalesMap = context.getApplicationCommandsContext().getBaseNameToLocalesMap();
	}

	@NotNull
	@Override
	public Map<DiscordLocale, String> apply(@NotNull String localizationKey) {
		final Map<DiscordLocale, String> map = new HashMap<>();

		baseNameToLocalesMap.forEach((baseName, locales) -> {
			for (Locale locale : locales) {
				final Localization instance = Localization.getInstance(baseName, locale);
				if (instance != null) {
					if (instance.getEffectiveLocale() != locale) {
						if (Logging.tryLog(baseName, locale.toLanguageTag(), instance.getEffectiveLocale().toLanguageTag())) {
							LOGGER.warn("Localization bundle '{}' with locale '{}' was specified to be valid but was not found, falling back to '{}'", baseName, locale, instance.getEffectiveLocale());
						}
					}

					final LocalizationTemplate template = instance.get(localizationKey);
					if (template != null) {
						map.put(DiscordLocale.from(locale), template.localize());
					} else if (DebugBuilder.isLogMissingLocalizationEnabled()) {
						if (Logging.tryLog(baseName, locale.toLanguageTag(), localizationKey)) {
							LOGGER.warn("Localization template '{}' could not be found in bundle '{}' with locale '{}' or below", localizationKey, baseName, locale);
						}
					}
				} else if (Logging.tryLog(baseName, locale.toLanguageTag())) {
					LOGGER.warn("Localization bundle '{}' with locale '{}' was specified to be valid but was not found.", baseName, locale);
				}
			}
		});

		return map;
	}
}

File: src/main/kotlin/com/freya02/botcommands/internal/commands/application/localization/BCLocalizationFunction.kt
package com.freya02.botcommands.internal.commands.application.localization

import com.freya02.botcommands.api.core.service.getService
import com.freya02.botcommands.api.localization.LocalizationService
import com.freya02.botcommands.internal.core.BContextImpl
import com.freya02.botcommands.internal.core.SingleLogger
import mu.KotlinLogging
import net.dv8tion.jda.api.interactions.DiscordLocale
import net.dv8tion.jda.api.interactions.commands.localization.LocalizationFunction
import java.util.*

private val logger = KotlinLogging.logger { }

internal class BCLocalizationFunction(private val context: BContextImpl) : LocalizationFunction {
    private val localizationService: LocalizationService = context.getService<LocalizationService>()
    private val baseNameToLocalesMap: Map<String, List<Locale>> = context.applicationConfig.baseNameToLocalesMap

    override fun apply(localizationKey: String): Map<DiscordLocale, String> {
        val map: MutableMap<DiscordLocale, String> = EnumMap(DiscordLocale::class.java)

        baseNameToLocalesMap.forEach { (baseName, locales) ->
            for (locale in locales) {
                val instance = localizationService.getInstance(baseName, locale)
                if (instance != null) {
                    if (instance.effectiveLocale !== locale) {
                        SingleLogger.current().tryLog(baseName, locale.toLanguageTag(), instance.effectiveLocale.toLanguageTag()) {
                            logger.warn(
                                "Localization bundle '{}' with locale '{}' was specified to be valid but was not found, falling back to '{}'",
                                baseName,
                                locale,
                                instance.effectiveLocale
                            )
                        }
                    }

                    val template = instance[localizationKey]
                    if (template != null) {
                        map[locale.toDiscordLocale()] = template.localize()
                    } else if (context.debugConfig.enabledMissingLocalizationLogs) {
                        SingleLogger.current().tryLog(baseName, locale.toLanguageTag(), localizationKey) {
                            logger.warn(
                                "Localization template '{}' could not be found in bundle '{}' with locale '{}' or below",
                                localizationKey,
                                baseName,
                                locale
                            )
                        }
                    }
                } else {
                    SingleLogger.current().tryLog(baseName, locale.toLanguageTag()) {
                        logger.warn("Localization bundle '{}' with locale '{}' was specified to be valid but was not found.", baseName, locale)
                    }
                }
            }
        }

        return map
    }

    private fun Locale.toDiscordLocale() = DiscordLocale.from(this)
}